connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/streaming/StreamingQuery.scala

```diff
@@ -0,0 +1,306 @@
+
+  /**
+   * Prints the physical plan to the console for debugging purposes.
+   * @since 3.5.0
+   */
+  def explain(): Unit
+
+  /**
+   * Prints the physical plan to the console for debugging purposes.
+   *
+   * @param extended whether to do extended explain or not
+   * @since 3.5.0
+   */
+  def explain(extended: Boolean): Unit
+}
+
+class RemoteStreamingQuery(
+  override val id: UUID,
+  override val runId: UUID,
+  override val name: String,
+  override val sparkSession: SparkSession) extends StreamingQuery {
+
+  override def isActive: Boolean = {
+    executeQueryCmd(_.setStatus(true)).getStatus.getIsActive
+  }
+
+  override def exception: Option[StreamingQueryException] = {
+    throw new UnsupportedOperationException("exception is not yet implemented")
+  }
+
+  override def status: StreamingQueryStatus = {
+    val statusResp = executeQueryCmd(_.setStatus(true)).getStatus
+    new StreamingQueryStatus(
+      message = statusResp.getStatusMessage,
+      isDataAvailable = statusResp.getIsDataAvailable,
+      isTriggerActive = statusResp.getIsTriggerActive
+    )
+  }
+
+  override def recentProgress: Array[StreamingQueryProgress] = {
+    executeQueryCmd(_.setRecentProgress(true))
+      .getRecentProgress
+      .getRecentProgressJsonList
+      .asScala
+      .map(json => new StreamingQueryProgress(json))
+      .toArray
+  }
+
+  override def lastProgress: StreamingQueryProgress = {
+    executeQueryCmd(_.setLastProgress(true))
+      .getRecentProgress
+      .getRecentProgressJsonList
+      .asScala
+      .headOption
+      .map(json => new StreamingQueryProgress(json))
+      .orNull
+  }
+
+  override def awaitTermination(): Unit = {
+    throw new UnsupportedOperationException("awaitTermination() is not yet implemented")
+  }
+
+  override def awaitTermination(timeoutMs: Long): Boolean = {
+    throw new UnsupportedOperationException("awaitTermination() is not yet implemented")
+  }
+
+  override def processAllAvailable(): Unit = {
+    executeQueryCmd(_.setProcessAllAvailable(true))
+  }
+
+  override def stop(): Unit = {
+    executeQueryCmd(_.setStop(true))
+  }
+
+  override def explain(): Unit = {
+    explain(extended = false)
+  }
+
+  override def explain(extended: Boolean): Unit = {
+    val explainCmd = StreamingQueryCommand
+      .ExplainCommand.newBuilder()
+      .setExtended(extended)
+      .build()
+
+    val explain = executeQueryCmd(_.setExplain(explainCmd))
+      .getExplain
+      .getResult
+
+    // scalastyle:off println
+    println(explain)
+    // scalastyle:on println
+  }
+
+  private def executeQueryCmd(
+    setCmdFn: StreamingQueryCommand.Builder => Unit // Sets the command field, like stop().
+  ): StreamingQueryCommandResult = {
+
+    val cmdBuilder = Command.newBuilder()
+    val queryCmdBuilder = cmdBuilder.getStreamingQueryCommandBuilder
+
+    // Set queryId.
+    queryCmdBuilder
+      .getQueryIdBuilder
+      .setId(id.toString)
+      .setRunId(runId.toString)
```

Comment:
Maybe we can make `id.toString` and `runId.toString` as `val` of `RemoteStreamingQuery`

Reply:
True, but this is not any noticeable overhead. `toString` is done only when we are sending an RPC.